prow/metrics/metrics.go

```diff
@@ -26,23 +26,59 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/push"
 	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/config"
 )
 
-// PushMetrics is meant to run in a goroutine and continuously push
+// NewPusher creates a Pusher which dynamically reads its config from the
+// ConfigAgent
+func NewPusher(ca configGetter) Pusher {
+	return Pusher{
+		configAgent: ca,
+	}
+}
+
+// Pusher is a struct that knows how to push metrics to prometheus
+type Pusher struct {
+	configAgent configGetter
+
+	// gatherer and waiter are only on this struct to be faked out in the tests
+	gatherer gathererFunc
+	waiter   func(time.Duration) <-chan time.Time
+}
+
+// Start is meant to run in a goroutine and continuously push
 // metrics to the provided endpoint.
-func PushMetrics(component, endpoint string, interval time.Duration) {
+func (pm Pusher) Start(component string) {
+	if pm.gatherer == nil {
+		pm.gatherer = push.FromGatherer
+	}
+	if pm.waiter == nil {
+		pm.waiter = time.After
+	}
+
 	sig := make(chan os.Signal, 1)
 	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
 
 	for {
+		interval := pm.configAgent.Config().PushGateway.Interval
+		waiter := pm.waiter(interval)
+
 		select {
-		case <-time.Tick(interval):
-			if err := push.FromGatherer(component, push.HostnameGroupingKey(), endpoint, prometheus.DefaultGatherer); err != nil {
-				logrus.WithField("component", component).WithError(err).Error("Failed to push metrics.")
+		case <-waiter:
+			endpoint := pm.configAgent.Config().PushGateway.Endpoint
```

Comment:
Why would we want to load config twice and potentially be in an inconsistent state between the two? Feels like we should load interval and endpoints atomically.

Reply:
See my earlier comment; we need to load here because otherwise the first push to the new endpoint is likely broken.

Reply:
> first push to the new endpoint is likely broken.

Assuming the endpoint has changed during the previous reload.

Reply:
Hmm, seems reasonable I guess